src/daft-plan/src/treenode.rs

```diff
@@ -55,3 +115,26 @@ impl DynTreeNode for PhysicalPlan {
         }
     }
 }
+
+/// Applies `f` to rewrite a `Arc<T>` without copying, if possible
+fn rewrite_arc<T, F: FnMut(T) -> DaftResult<Transformed<T>>>(
+    plan: Arc<T>,
+    mut f: F,
+) -> DaftResult<Transformed<Arc<T>>>
+where
+    T: Clone,
+{
+    f(unwrap_arc(plan))?.map_data(|new_plan| Ok(Arc::new(new_plan)))
+}
+
+/// Converts a `Arc<T>` without copying, if possible. Copies the plan
+/// if there is a shared reference
+pub fn unwrap_arc<T>(plan: Arc<T>) -> T
```

Comment:
curious on the reason we can't work on the Arc directly

Reply:
see https://github.com/Eventual-Inc/Daft/pull/2619#discussion_r1705970230